File: Week_02/java/src/_347_TopKFrequentElements.java
import java.util.Comparator;
import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;
/**
 * 第一遍：2020/05/30周六 ✅
 * 第二遍：2020/05/31周日
 * 第三遍：2020/06/06周六
 * 第四遍：2020/06/13周六
 */
class _347_TopKFrequentElements {
    /**
     * 使用一个堆来解决问题
     * 时间：O(nlogk) 空间：O(n)
     */
    public int[] topKFrequent(int[] nums, int k) {
        if (nums == null || nums.length == 0 || k <= 0) {
            return new int[0];
        }

        HashMap<Integer, Integer> map = new HashMap<>();
        for (int num : nums) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }

        PriorityQueue<Map.Entry<Integer, Integer>> queue = new PriorityQueue<>(Comparator.comparingInt(Map.Entry::getValue));
        for (Map.Entry<Integer, Integer> entry : map.entrySet()) {
            if (queue.size() < k) {
                queue.offer(entry);
            } else {
                if (queue.peek().getValue() < entry.getValue()) {
                    queue.poll();
                    queue.offer(entry);
                }
            }
        }

        int[] ret = new int[queue.size()];
        int idx = 0;
        for (Map.Entry<Integer, Integer> entry : queue) {
            ret[idx++] = entry.getKey();
        }
        return ret;
    }
}

File: Week_02/kotlin/src/_347_TopKFrequentElements.kt
import java.util.*
import kotlin.collections.HashMap

class _347_Heap {
    fun topKFrequent(nums: IntArray, k: Int): IntArray {
        val map = HashMap<Int, Int>()
        nums.forEach {
            map[it] = map.getOrDefault(it, 0) + 1
        }
        val heap = PriorityQueue<Map.Entry<Int, Int>> { l, r -> l.value - r.value }
        map.forEach {
            if (heap.size < k) {
                heap.offer(it)
            } else {
                if (heap.peek().value < it.value) {
                    heap.poll()
                    heap.offer(it)
                }
            }
        }
//        val ret = IntArray(k)
//        repeat(k) { ret[it] = heap.poll().key }
//        return ret
        return heap.toList().map { it.key }.toIntArray()
    }
}